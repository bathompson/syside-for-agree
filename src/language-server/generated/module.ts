/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

/******************************************************************************
 * This file was generated by langium-cli 0.5.0.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/

import { LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumSharedServices, LangiumServices, LanguageMetaData, Module, IParserConfig } from 'langium';
import { SysMlAstReflection } from './ast';
import { KerMLGrammar, SysMLGrammar } from './grammar';

export const KerMLLanguageMetaData: LanguageMetaData = {
    languageId: 'kerml',
    fileExtensions: ['.kerml'],
    caseInsensitive: false
};

export const SysMLLanguageMetaData: LanguageMetaData = {
    languageId: 'sysml',
    fileExtensions: ['.sysml'],
    caseInsensitive: false
};

export const parserConfig: IParserConfig = {
    recoveryEnabled: true,
    nodeLocationTracking: 'none',
};

export const SysMlGeneratedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices> = {
    AstReflection: () => new SysMlAstReflection()
};

export const KerMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => KerMLGrammar(),
    LanguageMetaData: () => KerMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};

export const SysMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => SysMLGrammar(),
    LanguageMetaData: () => SysMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};
