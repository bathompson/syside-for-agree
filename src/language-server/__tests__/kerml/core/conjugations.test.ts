/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

import { formatString } from "typescript-string-operations";
import { withQualifiedName, qualifiedTypeReference, anything } from "../../../../testing";
import { Conjugation, Documentation } from "../../../generated/ast";

const Common = `
type Original {
    in feature Input;
}
type Conjugate1;
type Conjugate2;
`;

test("conjugation can be parsed", async () => {
    return expect(
        Common +
            `
    conjugation c1 conjugate Conjugate1 conjugates Original;
    conjugation c2 conjugate Conjugate2 ~ Original {
        doc /* same as c1 */
    }`
    ).toParseKerML({
        relationships: [
            {
                $type: Conjugation,
                ...withQualifiedName("c1"),
                specific: qualifiedTypeReference("Conjugate1"),
                general: qualifiedTypeReference("Original"),
            },
            {
                $type: Conjugation,
                ...withQualifiedName("c2"),
                specific: qualifiedTypeReference("Conjugate2"),
                general: qualifiedTypeReference("Original"),
                docs: [
                    {
                        $type: Documentation,
                        body: "/* same as c1 */",
                    },
                ],
            },
        ],
    });
});

test.concurrent.each(["conjugates", "~"])(
    "conjugation may be omitted without identifiers with '%s'",
    async (token: string) => {
        return expect(
            formatString(
                Common +
                    `
    conjugate Conjugate1 {0} Original;`,
                token
            )
        ).toParseKerML({
            relationships: [
                {
                    $type: Conjugation,
                    specific: qualifiedTypeReference("Conjugate1"),
                    general: qualifiedTypeReference("Original"),
                },
            ],
        });
    }
);

test.concurrent.each(["conjugates", "~"])(
    "type can declare owned conjugations with '%s'",
    async (token: string) => {
        return expect(
            formatString(
                `
    type Original;
    type Conjugate1 {0} Original;`,
                token
            )
        ).toParseKerML({
            elements: [
                ...anything(1),
                {
                    ...withQualifiedName("Conjugate1"),
                    conjugates: [qualifiedTypeReference("Original")],
                },
            ],
        });
    }
);
