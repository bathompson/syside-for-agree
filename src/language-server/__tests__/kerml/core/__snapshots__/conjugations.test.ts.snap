// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conjugation can be parsed: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Original",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [
            {
              "$type": "FeatureMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "Input",
                "direction": "in",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate1",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate2",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Conjugation",
          "annotations": [],
          "chains": [],
          "declaredName": "c1",
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Original",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [
                    {
                      "$type": "FeatureMembership",
                      "annotations": [],
                      "chains": [],
                      "element": {
                        "$type": "Feature",
                        "aliases": [],
                        "annotatingMembers": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "filters": [],
                        "imports": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "members": [],
                        "namespaceMembers": [],
                        "prefixes": [],
                        "relationshipMembers": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                      "prefixes": [],
                    },
                  ],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Conjugate1",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Conjugation",
          "annotations": [
            {
              "$type": "Annotation",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* same as c1 */",
                "prefixes": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "chains": [],
          "declaredName": "c2",
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Original",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [
                    {
                      "$type": "FeatureMembership",
                      "annotations": [],
                      "chains": [],
                      "element": {
                        "$type": "Feature",
                        "aliases": [],
                        "annotatingMembers": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "filters": [],
                        "imports": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "members": [],
                        "namespaceMembers": [],
                        "prefixes": [],
                        "relationshipMembers": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                      "prefixes": [],
                    },
                  ],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Conjugate2",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
    ],
  },
}
`;

exports[`conjugation may be omitted without identifiers with '~': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Original",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [
            {
              "$type": "FeatureMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "Input",
                "direction": "in",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate1",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate2",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Conjugation",
          "annotations": [],
          "chains": [],
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Original",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [
                    {
                      "$type": "FeatureMembership",
                      "annotations": [],
                      "chains": [],
                      "element": {
                        "$type": "Feature",
                        "aliases": [],
                        "annotatingMembers": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "filters": [],
                        "imports": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "members": [],
                        "namespaceMembers": [],
                        "prefixes": [],
                        "relationshipMembers": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                      "prefixes": [],
                    },
                  ],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Conjugate1",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
    ],
  },
}
`;

exports[`conjugation may be omitted without identifiers with 'conjugates': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Original",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [
            {
              "$type": "FeatureMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "Input",
                "direction": "in",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate1",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate2",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Conjugation",
          "annotations": [],
          "chains": [],
          "elements": [],
          "prefixes": [],
          "reference": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Original",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [
                    {
                      "$type": "FeatureMembership",
                      "annotations": [],
                      "chains": [],
                      "element": {
                        "$type": "Feature",
                        "aliases": [],
                        "annotatingMembers": [],
                        "declaredName": "Input",
                        "direction": "in",
                        "filters": [],
                        "imports": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "members": [],
                        "namespaceMembers": [],
                        "prefixes": [],
                        "relationshipMembers": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                      "prefixes": [],
                    },
                  ],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "TypeReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Class",
                  "aliases": [],
                  "annotatingMembers": [],
                  "declaredName": "Conjugate1",
                  "filters": [],
                  "imports": [],
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
        "prefixes": [],
      },
    ],
  },
}
`;

exports[`type can declare owned conjugations with '~': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Original",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate1",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [
            {
              "$type": "Conjugation",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "Original",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`type can declare owned conjugations with 'conjugates': toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Original",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Class",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Conjugate1",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [
            {
              "$type": "Conjugation",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Class",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "Original",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;
